src/zjs_fs.c

```diff
@@ -0,0 +1,940 @@
+
+        return ZJS_UNDEFINED;
+    } else {
+        return jerry_create_number(ret);
+    }
+}
+
+static jerry_value_t zjs_read_sync(const jerry_value_t function_obj,
+                                   const jerry_value_t this,
+                                   const jerry_value_t argv[],
+                                   const jerry_length_t argc) {
+    return zjs_read(function_obj, this, argv, argc, 0);
+}
+
+static jerry_value_t zjs_read_async(const jerry_value_t function_obj,
+                                    const jerry_value_t this,
+                                    const jerry_value_t argv[],
+                                    const jerry_length_t argc) {
+    return zjs_read(function_obj, this, argv, argc, 1);
+}
+
+static jerry_value_t zjs_write(const jerry_value_t function_obj,
+                               const jerry_value_t this,
+                               const jerry_value_t argv[],
+                               const jerry_length_t argc,
+                               uint8_t async)
+{
+    file_handle* handle;
+    int position = 0;
+    jerry_value_t js_cb;
+
+    if (argc < 5) {
+        ERR_PRINT("not enough parameters\n");
+        return ZJS_UNDEFINED;
+    }
+
+    if (!jerry_value_is_object(argv[0])) {
+        ERR_PRINT("first parameter was not file descriptor\n");
+        return ZJS_UNDEFINED;
+    }
+    if (!jerry_value_is_object(argv[1])) {
+        ERR_PRINT("second parameter was not buffer\n");
+        return ZJS_UNDEFINED;
+    }
+    if (!jerry_value_is_number(argv[2])) {
+        ERR_PRINT("third parameter was not offset\n");
+        return ZJS_UNDEFINED;
+    }
+    if (!jerry_value_is_number(argv[3])) {
+        ERR_PRINT("fourth parameter was not length\n");
+        return ZJS_UNDEFINED;
+    }
+    if (argc == 6) {
+        if (!jerry_value_is_number(argv[4])) {
+            ERR_PRINT("position parameter was not number\n");
+            return ZJS_UNDEFINED;
+        }
+        if (async) {
+            if (!jerry_value_is_function(argv[5])) {
+                ERR_PRINT("sixth parameter was not callback function\n");
+                return ZJS_UNDEFINED;
+            }
+            js_cb = argv[5];
+        }
+        position = jerry_get_number_value(argv[4]);
+    } else {
+        if (async) {
+            if (!jerry_value_is_function(argv[4])) {
+                ERR_PRINT("fifth parameter was not callback function\n");
+                return ZJS_UNDEFINED;
+            }
+            js_cb = argv[4];
+        }
+    }
+    if (!jerry_get_object_native_handle(argv[0], (uintptr_t*)&handle)) {
+        ERR_PRINT("native handle not found\n");
+        return ZJS_UNDEFINED;
+    }
+    zjs_buffer_t* buffer = zjs_buffer_find(argv[1]);
+    uint32_t offset = jerry_get_number_value(argv[2]);
+    uint32_t length = jerry_get_number_value(argv[3]);
+
+    // if a position was specified, seek to it before writing
+    if (fs_seek(&handle->fp, position, SEEK_SET) != 0) {
```

Comment:
True, you need to make sure you have room in the target buffer since you are using its internal C buffer directly. We might want a helper function added to Buffer to assist, see what's needed.